### PR TITLE
Migrate to sbt 0.13.16

### PR DESCRIPTION
**Pull request by dwijnand**
I was passing by, seeing if I find a nice solution to #423, and ended up upgrading the build instead.

Feedback welcome - I'm happy to dial back some opinions :)

### CODE REVIEW

**Comment by havocp** [comment]
Thanks! This looks generally good to me, but I haven't been staying informed on sbt changes; if @eed3si9n or others more in the loop want to review that would be welcome. I'll plan to merge after a bit if nobody has any comments.

File: build.sbt [hunk 1]
```diff
@@ -13,10 +13,70 @@ scalacOptions in GlobalScope in Test := Seq("-unchecked", "-deprecation", "-feat
...
 
+val sonatype = new PublishToSonatype {
+  def projectUrl    = "https://github.com/typesafehub/config"
+  def developerId   = "havocp"
+  def developerName = "Havoc Pennington"
+  def developerUrl  = "http://ometer.com/"
+  def scmUrl        = "git://github.com/typesafehub/config.git"
+}
+
+lazy val commonSettings: Seq[Setting[_]] = Def settings (
```
Comment (eed3si9n) [inline build.sbt]: `Def.settings` definitely.

File: build.sbt [hunk 2]
```diff
@@ -13,10 +13,70 @@ scalacOptions in GlobalScope in Test := Seq("-unchecked", "-deprecation", "-feat
...
+  unpublished,
+  javaVersionPrefix in javaVersionCheck := None
+)
+
+lazy val root = project in file(".") settings (
```
Comment (eed3si9n) [inline build.sbt]:
```scala
(project in file("."))
  .settings(...)
```

File: build.sbt [hunk 3]
```diff
@@ -13,10 +13,70 @@ scalacOptions in GlobalScope in Test := Seq("-unchecked", "-deprecation", "-feat
...
+  packageDoc := (packageDoc in (configLib, Compile)).value,
+  aggregate in checkstyle := false,
+  checkstyle := (checkstyle in (configLib, Compile)).value
+) aggregate(
+  testLib, configLib,
+  simpleLibScala, simpleAppScala, complexAppScala,
+  simpleLibJava, simpleAppJava, complexAppJava
+)
+
+lazy val configLib =  Project("config", file("config")) settings (
```
Comment (eed3si9n) [inline build.sbt]:
```scala
.settings(...)
```

File: build.sbt [hunk 4]
```diff
@@ -13,10 +13,70 @@ scalacOptions in GlobalScope in Test := Seq("-unchecked", "-deprecation", "-feat
...
+  sonatype.settings,
+  osgiSettings,
+  OsgiKeys.exportPackage := Seq("com.typesafe.config", "com.typesafe.config.impl"),
+  publish := sys.error("use publishSigned instead of plain publish"),
+  publishLocal := sys.error("use publishLocalSigned instead of plain publishLocal")
+) enablePlugins SbtOsgi dependsOn testLib % "test->test"
```
Comment (eed3si9n) [inline build.sbt]:
```scala
  .enablePlugins(...)
  .dependsOn(...)
```

File: project/PublishToSonatype.scala [hunk 5]
```diff
@@ -0,0 +1,48 @@
+import sbt._, Keys._
+
+// from https://raw.github.com/paulp/scala-improving/master/project/PublishToSonatype.scala
+abstract class PublishToSonatype {
```
Comment (eed3si9n) [inline project/PublishToSonatype.scala]: Why an abstract class here?
Shouldn't it be an object or an AutoPlugin?

Comment (dwijnand) [inline project/PublishToSonatype.scala]: I has undefined defs.

**Comment by dwijnand** [comment]
Who defines sbt's idioms, you infix-hater :P (you requested changes coming right up)

**Pull request merged**